`IndsNetwork` additions

leburgel opened the pull request:
Some changes and additions for `IndsNetwork`s:

- Defined `is_directed` for `AbstractIndsNetwork` subtypes so they work with `Graphs.SimpleTraits`
- Added a `uniqueinds` method similar to the one for `AbstractITensorNetwork`.
- Added a `merge` method for merging `AbstractIndsNetwork`s, which just corresponds to a set union of all site and link indices of the network. Proved quite useful for me, though naming and usefulness in general might be debatable.
- Extended the `IndsNetwork` constructor to allow specifying the indices for every site/link of the network individually by means of a `Dictionary`, as well as specifying multiple indices per site/link.
- Extended the `siteind` system for `IndsNetwork`s to allow passing either an `Integer`, `String` or `Function` sitetype, as well as specifying the sitetype for every site individually. Now also allows passing any `Graphs.AbstractGraph` subtype, solving #7.
- Reorganized tests to use separate files.

mtfishman commented on src/abstractindsnetwork.jl:
```diff
@@ -4,6 +4,45 @@ abstract type AbstractIndsNetwork{I} <:
+# Convenience functions
+# 
+
+function Base.merge(is1::AbstractIndsNetwork{I}, is2::AbstractIndsNetwork{I}) where {I}
```
The closest Graphs.jl function for this seems to be [`Graphs.union`](https://juliagraphs.org/Graphs.jl/dev/core_functions/operators/#Base.union-Union{Tuple{T},%20Tuple{T,%20T}}%20where%20T%3C:Graphs.SimpleGraphs.AbstractSimpleGraph).

So this could be implemented as a more generic function, where we overload `Graphs.union` on a `DataGraph` but then allow users to pass a custom function for how to merge the metadata. Then this functionality could be implemented through `Graphs.union` by passing `unioninds` as the custom metadata merging function. We don't need to implement that in this PR, I'm just trying to think about what the more general principle or functionality behind this function might be since that might inspire a principled name to use here (i.e. the design in terms of `Graphs.union` might lead to calling this function as `union(unioninds, is1, is2)`, for example).

leburgel replied:
Sounds good. For now I can just rename it to specifically overload `Graphs.union` for `IndsNetwork`s here, which can then be replaced by a more general interface for once it is implemented for `DataGraph`s? Or would you already like to add the custom metadata merging function?

mtfishman replied:
Maybe we should name it something more specific for now like `union_all_inds` and think about the more general interface in a future PR.

mtfishman commented:
Thanks @leburgel, this looks good to me. It looks like there are some merge conflicts caused by #13 but once those are fixed I'll merge this.

Gotta get the tests working on this package, it's tough to figure out what's going on when dealing with unregistered packages...

leburgel commented:
Conflicts are resolved and tests pass locally, no idea how to fix the actual CI though.

mtfishman commented:
In #13 I deleted the file `examples/peps/utils.jl` since I moved it all into the library, but it looks like you added it back during your merge.

mtfishman commented:
(I think I'll just have to bite the bullet and register all of the dependent packages to get CI working better.)

leburgel commented:
> In #13 I deleted the file `examples/peps/utils.jl` since I moved it all into the library, but it looks like you added it back during your merge.

My bad, removed it again.

mtfishman commented:
Thanks again for the contribution!
